events/include/gz/common/Event.hh

```diff
@@ -55,8 +67,54 @@ namespace gz
       /// \param[in] _sig True if the event has been signaled.
       public: void SetSignaled(bool _sig);
 
+      /// \internal
+      /// \brief Removes queued connections.
+      /// We assume that this function is called from a Signal function.
+      protected: void Cleanup();
+
       /// \brief True if the event has been signaled.
       private: bool signaled;
+
+      /// \brief A private helper class used in maintaining connections.
+      protected: class EventConnection
+      {
+        /// \brief Constructor
+       public: EventConnection(bool _on, const std::any &_cb,
+                        const ConnectionPtr &_publicConn)
+            : callback(_cb), publicConnection(_publicConn)
+        {
+          // Windows Visual Studio 2012 does not have atomic_bool constructor,
+          // so we have to set "on" using operator=
+          this->on = _on;
+        }
+
+        /// \brief On/off value for the event callback
+        public: std::atomic_bool on;
+
+        /// \brief Callback function
+        public: std::any callback;
+
+        /// \brief A weak pointer to the Connection pointer returned by Connect.
+        /// This is used to clear the Connection's Event pointer during
+        /// destruction of an Event.
+        public: std::weak_ptr<Connection> publicConnection;
+      };
+IGN_UTILS_WARN_IGNORE__DLL_INTERFACE_MISSING
+      /// \def EvtConnectionMap
+      /// \brief Event Connection map typedef.
+      protected: typedef std::map<int, std::unique_ptr<EventConnection>>
+                    EvtConnectionMap;
```

Comment:
Prefer `using`.  I think we can probably switch to `unordered_map` here as well?

```suggestion
      protected: using EvtConnectionMap = std::map<int, std::unique_ptr<EventConnection>>;
```